torch.nn.MultiheadAttention fix

Hahu803 opened the pull request:
Differential Revision: D34321018

Hahu803 commented on torch/nn/modules/activation.py:
```diff
@@ -1038,7 +1038,16 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
         is_batched = query.dim() == 3
         if self.batch_first and is_batched:
-            query, key, value = [x.transpose(1, 0) for x in (query, key, value)]
+            if key is value:
```
This will make a difference in the _in_projection_packed method in torch/nn/functional.py. When we make transpose for q,k,v respectively, it will never satisfy the conditions of self-attention (needing "key is value" and "query is key") or encoder-decoder attention (needing "key is value" and "not query is key"). It's true it should not affect the values. However we observed if we need 3 linear module instantiations, it can take more time in the forward/backward processes (especially when we use mix-precision training, the effect of extra casting also slows down the training). We fix this so that in the self-attention (our case), only one linear module is needed and thus speed up out training, which has been proved in out testing (posted in the internal diff)

Hahu803 replied:
BTW, I am a little confused about the test failure. Seems like it fails in test_adam and some error msg like:
test_fixed_cuda_assert_async (__main__.TestCuda) ... C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\cuda\TensorCompare.cu:144: block: [0,0,0], thread: [0,0,0] Assertion `input[0] != 0` failed.
2022-02-18T02:48:23.8845865Z Traceback (most recent call last):
2022-02-18T02:48:23.8847495Z   File "<string>", line 4, in <module>
2022-02-18T02:48:23.8848545Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\cuda\__init__.py", line 495, in synchronize
2022-02-18T02:48:23.8849408Z     return torch._C._cuda_synchronize()
2022-02-18T02:48:23.8850077Z RuntimeError: CUDA error: device-side assert triggered

Full logs are seen in https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/1769331/signedlogcontent/27?urlExpires=2022-02-18T06%3A04%3A12.7207992Z&urlSigningMethod=HMACV1&urlSignature=gev8b7dbjy9NngHpNRP%2BqHFdUO0FYESAhrkpFWxTY2s%3D. I don't see the relationship between my change and these errors. Could you give any hints about this? Thanks.

albanD replied:
Thanks for the details.
I think that you want to update the comments in the code to reflect the fact that this whole logic aims at preserving the "is" property between the three values. And it is not really important what each case corresponds to,

Hahu803 replied:
Yeah thanks for the great suggestion! Let me update the comments to avoid confusion.

Commit:
[pytorch][nn] torch.nn.MultiheadAttention fix (#73053)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73053

We found we cannot hit the `self-attention` case (where q,k,v are the same) in https://fburl.com/code/ebgri4k5 when we have the flag "`batch_first=True`" in the input of torch.nn.MultiheadAttention. Since even when we have the same q,k,v, we need to do transpose op for them separately if "`batch_first=True`"  in https://fburl.com/code/h26mlbti. After this op, q,k,v will be different objects and the conditions "`k is v` or `q is k`" will be broken.

We need to adjust the transpose strategy to make sure that it aligns with the logic in https://fburl.com/code/0mg6i4yn and we can still satisfy the `self-attention`  and `encoder-decoder attention` conditions when "`batch_first=True`".

Test Plan:
Use ads Transformer 1x model (where q,k,v are the same) as example and run:
```
CUDA_VISIBLE_DEVICES=7 buck run mode/opt -c fbcode.nvcc_arch=a100 -c fbcode.enable_nccl_a2a=1 //hpc/models/ads:ads_transformer_1x_2021h1_launcher -- +launcher=local launcher.num_trainers=1 +data_loader=random +mode=mast model.shrink=true model.max_ind_range=1000 data_loader.num_batches=2500 profiling_trace=true
```
After this fix and given the input of multiheadAttention of shape (`batch_size=512, length=80, emb_dim=512`), we improve QPS from 4705 to 4937. This is because we save the unnecessary extra linear module instantiation in forward/backward when q,k,v are the same and `batch_first=true`. From the trace examples, we can see the difference:

Before fixing (see [trace example](https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/hpc.huaqingxiong.local/rank-0/rank-0.Feb_17_16_10_15.1653768.trace.gz&bucket=hpc_traces)) and the key part in the following screenshot:
{F702341108}

After fixing (see [trace example](https://our.intern.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/hpc.huaqingxiong.local/rank-0/rank-0.Feb_17_15_59_03.1533566.trace.gz&bucket=hpc_traces))

{F702343574}

Differential Revision: D34321018

fbshipit-source-id: aa4a4877c4314512e05eed858b5849274d39675c

albanD reviewed (approved):
SGTM
